scripts/upgrade

```diff
@@ -1,153 +1,173 @@
-    fi
-    
-    ynh_app_setting_delete $app skipped_uris
-    ynh_sso_access "/index.php?r=admin,/index.php?r=plugins,/scripts"
+# If datadir doesn't exist, create it
+if [ -z "$datadir" ]; then
+   datadir=/home/yunohost.app/$app/upload
+   ynh_app_setting_set --app=$app --key=datadir --value=$datadir
+   mkdir -p $datadir
+    # Remove upload if empty
+    rmdir $datadir && mv "$final_path/upload" $datadir || true
+   chown -R $app:www-data "$datadir"
+   chmod -R o-rwx "$datadir"
+fi
 
-    ynh_configure migrations/2.0.5.sql ./2.0.5.sql
-    mysql -u $app -p$db_pwd $db_name < ./2.0.5.sql
+# Cleaning legacy permissions
+if ynh_legacy_permissions_exists; then
+    ynh_legacy_permissions_delete_all
+
+    ynh_app_setting_delete --app=$app --key=is_public
 fi
-if ynh_version_le "2.62.2-1" ; then
-    is_public=${is_public:-0}
-    prefix=${prefix:-prefix_}
-    db_name=${db_name:-$app}
-    final_path=$(ynh_app_setting_get $app local_path)
-    ynh_save_args path_url is_public prefix db_name final_path db_user
+
+if ! ynh_permission_exists --permission="admin"; then
+    # Create the required permissions
+    ynh_permission_create --permission="admin" --url="/admin" --allowed=$admin
 fi
-if ynh_version_le "2.62.2-2" ; then
-    # Move old templates  and themes in backup dir
-    mv $final_path/upload/templates /home/yunohost.backup/$app.bkp/old-templates
-    mkdir -p /home/yunohost.backup/$app.bkp/old-themes
-    mv $final_path/themes/* /home/yunohost.backup/$app.bkp/old-themes/
+
+# Create a permission if needed
+if ! ynh_permission_exists --permission="api"; then
+    ynh_permission_create --permission="api" --url="/api" --allowed="visitors" --show_tile="false" --protected="true"
 fi
 
+#=================================================
+# CREATE DEDICATED USER
+#=================================================
+ynh_script_progression --message="Making sure dedicated system user exists..." --weight=1
+
+# Create a dedicated user (if not existing)
+ynh_system_user_create --username=$app --home_dir="$final_path"
+
 #=================================================
 # STANDARD UPGRADE STEPS
 #=================================================
 # DOWNLOAD, CHECK AND UNPACK SOURCE
 #=================================================
 
-# Backup config, plugins and themes and delete previous files
-ynh_keep_if_no_upgrade plugins/*
-ynh_keep application/config/config.php
-ynh_keep upload
+if [ "$upgrade_type" == "UPGRADE_APP" ]
+then
+    ynh_script_progression --message="Upgrading source files..." --weight=1
 
-# Copie new files and restore config, plugins, upload and themes
-ynh_secure_remove $final_path
-ynh_setup_source "$final_path"	# Télécharge la source, décompresse et copie dans $final_path
+    # Download, check integrity, uncompress and patch the source from app.src
+    ynh_setup_source --dest_dir="$final_path" --keep="$final_path/plugins/* $final_path/application/config/config.php $final_path/upload"
```

Comment:
Does `--keep` really works with wildcard ? Why not just `$final_path/plugins/` ?

Reply:
Cause we want new plugins...
```suggestion
    ynh_setup_source --dest_dir="$final_path" --keep="$(echo $final_path/plugins/*) $final_path/application/config/config.php $final_path/upload"
```